### ISSUE TITLE
Aim of this package

### ISSUE THREAD

**Issue by Vlix** [cdepillabout/password#7]
I'd say a package that just delivers plug-n-go password functionality would be very helpful to the Haskell ecosystem, since the specific fine-grained functionality is already provided by the specific packages (i.e.. `scrypt`, `bcrypt`, `cryptonite`, etc.)

So making this package as straight-forward and a complete no-brainer in its use would be a good idea, in my honest opinion.
This would entail not having any functions in which you can provide your own salt, no parameters to be set. Just have everything be the default industry standard. This way the package can be a great asset to anyone who only wants to handle passwords and doesn't want to think about if they're doing it the right way.
If someone knows what they're doing, we can point them to the library where they can adjust parameters and salts and whatever.

What would you think about this goal for the `password` package?

---

Also, side-note, I've just seen that the `scrypt` library hasn't been updated since 2014, so it might be a good idea to switch to `cryptonite`, since it is kept up to date and also has all the other (`Bcrypt`, `PBKDF2`) password algorithms.

**Comment by cdepillabout** [comment]
Thanks for creating this issue.

> What would you think about this goal for the `password` package?

Originally this package had two goals:

1. Create canonical `Pass` and `PassHash` types that can be shared throughout the Haskell ecosystem.  (Although now that I think about this, I should have just created the `password-types` package and only included these two datatypes.)

    There are various password-related packages on Hackage, and they all define their own types representing passwords and password hashes (and some of them just use the stringly-typed `ByteString` for both).  I wanted to create a shared vocabulary for talking about this.

2. Add convenience instances for `Pass` and `PassHash` for common libraries like `aeson` and `persistent`.

    When writing a webapps in Haskell, I often had to create a `Pass` and `PassHash` type locally in order to write a `FromJSON` instance (and similar instance for `persistent`).

    I got tired of this and decided to just create a library to share this functionality.

It wasn't necessarily one of my goals when I started, but it also makes sense to actually provide the password hashing functionality (and password hash checking functionality) in this library for ease-of-use.

> So making this package as straight-forward and a complete no-brainer in its use would be a good idea, in my honest opinion... If someone knows what they're doing, we can point them to the library where they can adjust parameters and salts and whatever.

In general I agree with this.

Ideally we would be able to support users when they decide they want additional parameters, features, etc.  `password` and `password-instances` should be *easy* to use in the normal case, and still *possible* to use in the case where a user wants to configure various settings.

Currently `password` doesn't expose any functions for changing settings (parameters for the underlying hash function).  I think  this is fine until someone explicitly requests it.  For now, I think we should be fine using industry-standard settings, as you suggest.

-------------------------------------------------------------

> Also, side-note, I've just seen that the scrypt library hasn't been updated since 2014, so it might be a good idea to switch to cryptonite, since it is kept up to date and also has all the other (Bcrypt, PBKDF2) password algorithms.

This sounds like a good idea!  If you (or anyone else) wanted to send a PR implementing this, I'd be happy to merge it.

I'm not sure if the actual format of the hashes produced by `scrypt` is standardized, but we'll have to be sure that `cryptonite` produces hashes of the same format.  If we updated to `cryptonite` and the format of the hashes changed, that could be a very surprising change to users of `password`!

As far as I know, there aren't any users of `password` other than me, so I'd be fine with doing this now.  However, I'd be much less willing to do this if `password` gets more users.

(If the format of the hashes doesn't change, then I'd be happy to update to `cryptonite` whenever.)

**Comment by Vlix** [comment]
Ok, I'd opt for changing to `cryptonite` then. And maybe keep tests that check that the `cryptonite` hashes are identical to the `scrypt` hashes.
I've tried to find the "default" hashing format for `scrypt`, but it doesn't really seem like there is one, so I guess just sticking to the `scrypt` library's format is fine.

I'd say, instead of the `hashWithSalt` function, maybe better to make one function in which you can just put ALL the parameters, where users can use `defaultParams {specificField = someValue} :: HashParams` to change one value, or just fill in their own `HashParams` to completely customize it, like:

```haskell
data ScryptParams = ScryptParams {
  salt :: Either Int ByteString, -- ^ Either give the length of a salt to be generated, or the salt itself
  rounds :: Int,
  blockSize :: Int,
  parallellism :: Int
}
```

Something like this. Where the `defaultParams` would be `ScryptParams (Left 16) 16 8 1`

With the no-brainer functions at the top of the modules, and then a big disclaimer before the customizable and unsafe functions to only use those if you know what you're doing. "If you have any doubt about your knowledge of these algorithms, then just use the default settings, please."

**Comment by Vlix** [comment]
Oh btw, I actually only found one other package that tries to do what this package does, and that's [Yesod.Auth.Util.PasswordStore](https://hackage.haskell.org/package/yesod-auth/docs/Yesod-Auth-Util-PasswordStore.html)
And that one uses PBKDF1 and provides PBKDF2 as well, and at the time (2013) that was ok, but now these are not the ones you'd generally want to use. (if popular opinion is anything to go by)

**Comment by cdepillabout** [comment]
Okay, the idea for the changes to `cryptonite` and the `hashWithSalt` sound good to me!

I'll mark this as `help wanted` if you (or anyone else) would be interested in sending a PR for this.

**Comment by Vlix** [comment]
I've started working on the proposed changes, and I got Scrypt in its own module using cryptonite's functions... just one thing, the hashes made by `cryptonite` and `scrypt` differ.

You can't use one on hashes from the other.

I've made an issue in the cryptonite repo to ask if they have any idea what might be going on. One of them is likely wrong, it seems.

**Comment by Vlix** [comment]
For completeness: I've asked in the `cryptonite` issues section, and I didn't read well enough, because `Crypto.KDF.Scrypt`'s parameters expect the full amount of rounds instead of the `2log(N)` amount of rounds, so this has been fixed in #8 and is now equivalent to what `password-1.0.0.0` hashed to.
